### PR TITLE
Tensor parallelism for Mixture of Experts

### PR DESCRIPTION
**Pull request by siddharth9820**
Note: This is in conjunction with this [PR](https://github.com/microsoft/DeepSpeed/pull/2074) on the deepspeed repo.

This PR adds tensor parallelism for non-experts. This combined with ZeRO-2 allows us to scale to roughly 2x larger base models than ZeRO-2. When tensor parallelism is enabled only for non-experts, there are duplicate tokens at each gate. It is important to drop the duplicates before they reach the experts, otherwise we run into convergence issues. In `megatron/mpu/mappings.py`, we provide new autograd functions that drop these tokens and gather them, namely `_DropTokens` and `_AllGatherFromModelParallelRegion`. 
In the current implementation, we drop tokens right before the AlltoAll and gather them right after the AlltoAll. These calls are done in the Deepspeed codebase.
 

Update: This PR now supports tensor parallelism for experts as well. This can be enabled by passing the `--enable-expert-tensor-parallelism` argument.

### CODE REVIEW

**Comment by siddharth9820** [comment]
Comparing loss curves with no tensor parallelism

image

File: megatron/mpu/mappings.py
Comment (awan-10) [inline megatron/mpu/mappings.py:191]: @siddharth9820 -- where is this ReduceScatter being used?

Comment (siddharth9820) [reply megatron/mpu/mappings.py:191]: @awan-10, for now nowhere. I can remove it if you want. I kept it there because it can potentially be used in place of the all-reduce at the end of tensor-parallel self-attention.

Comment (awan-10) [reply megatron/mpu/mappings.py:191]: Got it. Good to remove from this PR but don't change things just yet. All of this code essentially should be inside deepspeed and not here. I can share more with you when we talk.

File: megatron/model/transformer.py
Comment (awan-10) [inline megatron/model/transformer.py:469]: moe_mp_size

**Comment by siddharth9820** [comment]
Here are the loss curves when tensor parallelism is used for both the experts and non experts.
image

File: megatron/mpu/layers.py
Comment (siddharth9820) [inline megatron/mpu/layers.py:247]: merge it into previous if

File: megatron/model/transformer.py
Comment (conglongli) [inline megatron/model/transformer.py:458]: This actually leads to an error since ep_size is not defined. After rolling back one commit (8398bb05b62249909117f069683da267a87cf960) the error disappeared and non-expert/expert MP work fine. Please make sure to fix this before merging. And still need full code review from @awan-10 .

**Review by conglongli — APPROVED** [review]

File: megatron/model/transformer.py
Comment (siddharth9820) [reply megatron/model/transformer.py]: fixed in the next commit

File: megatron/mpu/layers.py
Comment (awan-10) [reply megatron/mpu/layers.py:247]: this is done? Please resolve if it is.

File: megatron/mpu/layers.py
Comment (awan-10) [inline megatron/mpu/layers.py:241]: cleanup the comment? or make a doc string to explain if that is the intent.

**Pull request merged**
